### PR TITLE
decorated public API with stability states

### LINKED ISSUE #961: decorate all functionality in public API to indicate stability state

### PR DESCRIPTION
**Pull request by gregcaporaso**
Fixes #961.

Does not cover public API in ``sequence`` and ``io``.

Open question: do we (and if so, how) decorate sub-classes that don't define any of their own methods such as ``EfficiencyWarning`` and ``OverrideError``?

### CODE REVIEW

**Comment by gregcaporaso** [comment]
This is ready for review, but I think we might want to review as a team to discuss how specific functionality is decorated (e.g., are some things ``experimental`` that should be ``deprecated``). 

cc @ebolyen @jairideout @anderspitman

**Comment by gregcaporaso** [comment]
I'm currently able to generate and compile the C code when I add the decorators, but I haven't been able to get them to actually run. Here's an example of a small change and the corresponding error message:

```bash
$ git diff skbio/alignment/_ssw_wrapper.pyx
diff --git a/skbio/alignment/_ssw_wrapper.pyx b/skbio/alignment/_ssw_wrapper.pyx
index da5bb3e..214a3ab 100644
--- a/skbio/alignment/_ssw_wrapper.pyx
+++ b/skbio/alignment/_ssw_wrapper.pyx
@@ -11,6 +11,7 @@ import numpy as np
 cimport numpy as cnp
 from skbio.alignment import Alignment
 from skbio.sequence import Protein, Sequence
+from skbio.util._decorator import experimental

 cdef extern from "_lib/ssw.h":

@@ -712,7 +713,7 @@ cdef class StripedSmithWaterman:
                 i += 1
         return py_list_matrix

-
+@experimental(as_of="0.4.0")
 def local_pairwise_align_ssw(sequence1, sequence2, constructor=Sequence,
                              **kwargs):
     """Align query and target sequences with Striped Smith-Waterman.
```

Then:

```bash
$ nosetests --with-doctest skbio/alignment/
E
======================================================================
ERROR: Failure: AttributeError (attribute '__doc__' of 'builtin_function_or_method' objects is not writable)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/caporaso/.virtualenvs/skbio/lib/python2.7/site-packages/nose/loader.py", line 420, in loadTestsFromName
    addr.filename, addr.module)
  File "/Users/caporaso/.virtualenvs/skbio/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/Users/caporaso/.virtualenvs/skbio/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/Users/caporaso/Dropbox/code/scikit-bio/skbio/__init__.py", line 14, in <module>
    import skbio.io  # noqa
  File "/Users/caporaso/Dropbox/code/scikit-bio/skbio/io/__init__.py", line 311, in <module>
    import_module('skbio.io.clustal')
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/caporaso/Dropbox/code/scikit-bio/skbio/io/clustal.py", line 120, in <module>
    from skbio.alignment import Alignment
  File "/Users/caporaso/Dropbox/code/scikit-bio/skbio/alignment/__init__.py", line 235, in <module>
    from skbio.alignment._ssw_wrapper import (
  File "skbio/alignment/_ssw_wrapper.pyx", line 716, in init skbio.alignment._ssw_wrapper (skbio/alignment/_ssw_wrapper.c:11816)
    @experimental(as_of="0.4.0")
  File "/Users/caporaso/Dropbox/code/scikit-bio/skbio/util/_decorator.py", line 191, in __call__
    func.__doc__ = self._update_docstring(func.__doc__, state_desc)
AttributeError: attribute '__doc__' of 'builtin_function_or_method' objects is not writable

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (errors=1)
```

**Comment by ebolyen** [comment]
It would seem Cython makes these builtins. I would recommend just pulling that function out of the `.pyx` file and not worrying about the internal class.

**Comment by gregcaporaso** [comment]
The only failure now is a result of multiple warnings coming from the distribution plotting code under python 3.4 (those functions now raise a deprecation warning). We should discuss this one in our meeting this morning.

File: skbio/alignment/_alignment.py [hunk 1]
```diff
@@ -1431,6 +1466,10 @@ class StockholmAlignment(Alignment):
     seq2          UCC--G-GGGA
     //
     """
+
+    @deprecated(as_of="0.4.0", until="0.4.1",
```
Comment (jairideout) [inline skbio/alignment/_alignment.py]: remove instead

File: skbio/tree/_trie.py [hunk 2]
```diff
@@ -177,24 +179,49 @@ class CompressedTrie(object):
 
     """
 
+    @deprecated(as_of="0.4.0", until="0.4.1",
```
Comment (jairideout) [inline skbio/tree/_trie.py]: put the text in a var to avoid duplication (in the future we'll have a better way of deprecating entire classes)

**Comment by jairideout** [comment]
Thanks @gregcaporaso!

**Pull request merged**